File: tools/ai-model-pipeline/runtime-test/story_setup.py
"""Bounded native story setup. No generic dismissal, combat UI, or action retries."""
import time

def require(value,message):
    if not value:raise RuntimeError(message)

def complete(state):
    signals=state.get('signals') or {};panels=[state.get('portrait') or {},state.get('questConfirm') or {}];fsms=state.get('presenterFsms') or []
    return (state.get('complete') is True and state.get('messagePresent') is False and state.get('messageType')=='None'
        and state.get('queueCount')==0 and state.get('currentContinuationPresent') is False
        and len(fsms)==3 and all(f.get('present') is True and f.get('enabled') is False for f in fsms)
        and signals.get('modalOpen') is False and signals.get('choiceOpen') is False and signals.get('warnings')==[]
        and all(p.get('activeSelf') is False and p.get('activeInHierarchy') is False and p.get('fullyOpened') is False for p in panels))

def scope(state):
    require(all(state.get(k) is True for k in ('inSession','livingOwnedParty','outsideCombat','outsideDungeon')),'Story setup requires living owned SP overworld outside both combat sessions/dungeon')
    require(all(type(state.get(k)) is int and state[k]!=0 for k in ('coordinatorId','presenterId','heroInstanceId')),'Story native owner IDs unavailable')
    require(state.get('status') not in ('unsupported-choice','unsupported-modal'),'Unknown modal/choice; no automatic story operation')
    return tuple(state[k] for k in ('root','coordinatorId','presenterId','heroInstanceId'))+(state.get('coreIdentity'),)

def submission(state):
    surface=state.get('actionableSurface');require(surface in ('portrait','questConfirm'),'No exact actionable story page')
    require(state.get('messageType')=='StoryQuestMessage' and state.get('messagePresent') is True and state.get('messageClosed') is False,'Unknown native story message')
    require(state.get('presenterMessageId')==state.get('messageId'),'Coordinator/presenter message mismatch')
    require(state.get('questPresent') is True and state.get('questRegisteredReference') is True,'Native presenter quest must be exact registered object')
    signals=state.get('signals') or {};require(signals.get('modalType')=='StoryQuestMessage' and signals.get('choiceOpen') is False,'Story signals changed')
    keys=('coordinatorId','messageId','presenterId','pageIndex','questId','heroInstanceId','phase','contentSha256');result={k:state[k] for k in keys};panel=state[surface]
    result.update(surface=surface,componentId=panel['componentId'],panelId=panel['panelId'],clickInstanceId=panel['focus']['instanceId'],clickContinuationId=panel['focus']['continuationId'])
    require(all(type(v) is int for k,v in result.items() if k not in ('surface','phase','contentSha256')),'Exact story integer identities required')
    require(result['messageId']>=0 and result['pageIndex']>=0 and result['questId']!=0 and result['componentId']!=0 and result['panelId']!=0,'Invalid story identity')
    return result

def page_key(request):return tuple(request[k] for k in ('coordinatorId','presenterId','heroInstanceId','messageId','pageIndex','phase','contentSha256','surface'))

def progressed(pending,state):
    if state.get('messagePresent') is False:return True
    current=state.get('messageId');page=state.get('pageIndex')
    require(type(current) is int and type(page) is int,'Missing native story progress IDs')
    if current!=pending['messageId']:return True
    phase=state.get('phase');content=state.get('contentSha256')
    if phase=='unavailable' and state.get('actionableSurface') is None:return False
    if phase==pending['phase'] and content==pending['contentSha256']:
        require(page>=pending['pageIndex'],'Native story page regressed without native phase advancement')
        return page>pending['pageIndex']
    if state.get('actionableSurface')=='questConfirm':
        return page>pending['pageIndex'] # Native final page increments before the confirmation path.
    if (pending['phase']=='DeliverMultiQuestMsgClosed' and phase=='DeliverSubQuestMsgClosed'
            and content!=pending['contentSha256'] and state.get('actionableSurface')=='portrait'):
        return True # Native rebuilds its messages and resets index for subquests.
    if state.get('actionableSurface') is None:return False
    raise RuntimeError('Unexpected native story phase/content transition')

def clear(runner):
    deadline=time.monotonic()+runner.a.wait_timeout;pin=None;pending=None
    submitted=set(getattr(runner,'story_page_keys',()));runner.story_page_keys=submitted
    clear_since=None;last_frame=-1;previous=None
    while time.monotonic()<deadline:
        state=runner.helper('story-state');runner.log('story-observation',state);current=scope(state)
        if pin is None:pin=current
        require(current==pin,'Story root/native owner/Core identity changed')
        require(type(state.get('frame')) is int and state['frame']>last_frame,'Fresh advancing story frame required');last_frame=state['frame']
        if pending is not None and progressed(pending,state):
            runner.log('story-native-progress',{'submission':pending,'observed':state});pending=None
        if complete(state):
            clear_since=time.monotonic() if clear_since is None else clear_since
            if time.monotonic()-clear_since>=2:
                runner.log('story-completion-observed',state);return state
        else:clear_since=None
        if state.get('actionableSurface') is not None and pending is None:
            if previous is not None:require(progressed(previous,state),'Native story phase/page did not advance')
            request=submission(state);key=page_key(request)
            require(key not in submitted and len(submitted)<64,'Story page already submitted or64-page bound exceeded')
            submitted.add(key);pending=request;previous=request # Preserve uncertainty before submitting exactly once.
            result=runner.helper('story-submit',request)
            require(result.get('status')=='native-story-page-submitted' and result.get('surface')==request['surface'],'Story submission outcome unverified; never retry')
            before=result.get('before') or {};require(submission(before)==request,'Native submission identity differs from observed request')
            runner.log('story-page-submitted',{'request':request,'result':result})
        time.sleep(.25)
    raise TimeoutError('Native story setup did not complete within budget; no action retried')

def staging_status(state):
    signals=state.get('signals') or {}
    if signals.get('modalOpen') is False and signals.get('choiceOpen') is False:return 'binding_metadata_observed'
    if signals.get('modalOpen') is True and signals.get('choiceOpen') is False and signals.get('modalType')=='StoryQuestMessage':return 'pending_story_message'
    raise RuntimeError('Unknown post-staging modal/choice; no automatic dismissal or combat action')

File: docs/evidence/entry-preparation-native-v1/reviewed-source/tools/ai-model-pipeline/runtime-test/StorySetup.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Reflection;
using System.Security.Cryptography;
using System.Text;
using UnityEngine;
using Newtonsoft.Json.Linq;

public sealed partial class RuntimeModelTest
{
    readonly HashSet<string> submittedStoryPages=new HashSet<string>();
    sealed class StoryContext
    {
        public MessageCoordinator coordinator;public MessagePresenter presenter;public CharacterOverworld cow;
        public FTKClickAnywhere click;public uiPortraitMessageHud portrait;public uiQuestConfirmHud confirm;public JObject view;
    }
    static JObject StoryPanel(Component hud,Component panel,bool fully,bool clickable,bool closeOnOkay)
    {
        return new JObject{{"componentId",hud==null?0:hud.GetInstanceID()},{"panelId",panel==null?0:panel.GetInstanceID()},
            {"hudActive",hud!=null && hud.gameObject.activeInHierarchy},{"activeSelf",panel!=null && panel.gameObject.activeSelf},
            {"activeInHierarchy",panel!=null && panel.gameObject.activeInHierarchy},{"fullyOpened",fully},{"clickable",clickable},{"closeOnOkay",closeOnOkay}};
    }
    static JObject StoryFocus(FTKClickAnywhere click,Component hud)
    {
        object continuation=click==null?null:typeof(FTKClickAnywhere).GetField("m_Continue",Members).GetValue(click);
        object callback=continuation==null?null:typeof(ContinueFSM).GetField("m_CompleteDelegate",Members).GetValue(continuation);
        bool exact=continuation!=null && hud!=null && StorySubmissionPolicy.ExactAction(callback,hud,"UseOkayButton") &&
            typeof(ContinueFSM).GetField("m_CompleteDelegate_1",Members).GetValue(continuation)==null &&
            typeof(ContinueFSM).GetField("m_FSM",Members).GetValue(continuation)==null &&
            typeof(ContinueFSM).GetField("m_ContinueObject",Members).GetValue(continuation)==null;
        ContinueFSM cfsm=continuation as ContinueFSM;
        object player=typeof(MessageCoordinator).Assembly.GetType("PhotonNetwork",true).GetProperty("player",Statics).GetValue(null,null);
        if(player==null)throw new InvalidOperationException("Native story input caller unavailable.");
        int localPlayerId=(int)player.GetType().GetProperty("ID",Members).GetValue(player,null);
        bool dispatch=cfsm!=null && cfsm.m_ID>=0 && cfsm.m_WaitClients==ContinueFSM.WaitClients.Self && cfsm.m_WaitCount==1 &&
            (cfsm.m_CallerID==localPlayerId || (localPlayerId==-1 && cfsm.m_IsLocal));
        return new JObject{{"instanceId",click==null?0:click.GetInstanceID()},{"active",click!=null && click.enabled && click.gameObject.activeInHierarchy},
            {"currentFocus",click!=null && FTKInput.Instance!=null && ReferenceEquals(FTKInput.Instance.m_CurrentInputFocus,click) && click.m_HasInputFocus},
            {"canClose",click!=null && (bool)typeof(FTKClickAnywhere).GetField("m_CanClose",Members).GetValue(click)},
            {"exactCallback",exact},{"localDispatchReady",dispatch},{"localPlayerId",localPlayerId},{"waitCount",cfsm==null?-1:cfsm.m_WaitCount},{"waitClients",cfsm==null?"unavailable":cfsm.m_WaitClients.ToString()},{"continuationId",cfsm==null?-1:cfsm.m_ID},{"continuationCallerId",cfsm==null?-1:cfsm.m_CallerID},
            {"callbackTargetId",callback is Action && ((Action)callback).Target is Component?((Component)((Action)callback).Target).GetInstanceID():0},
            {"callbackMethod",callback is Action?((Action)callback).Method.Name:null}};
    }
    static string StoryContentHash(MessagePresenter presenter)
    {
        if(presenter.m_Messages==null)return null;
        if(presenter.m_Messages.Count>256)throw new InvalidOperationException("Native story content exceeds256 pages.");
        JArray entries=new JArray();int chars=0;
        foreach(MessagePresenter.PresentEntry entry in presenter.m_Messages)
        {
            if(entry==null)throw new InvalidOperationException("Missing native story page.");
            chars+=(entry.m_Message??"").Length+(entry.m_UserNPC??"").Length;
            if(chars>262144)throw new InvalidOperationException("Native story content exceeds bounded262144 characters.");
            entries.Add(new JObject{{"talker",(int)entry.m_Talker},{"userNpc",entry.m_UserNPC},{"message",entry.m_Message}});
        }
        using(SHA256 hash=SHA256.Create())return BitConverter.ToString(hash.ComputeHash(Encoding.UTF8.GetBytes(entries.ToString()))).Replace("-","").ToLowerInvariant();
    }
    static string StoryPhase(StoryContext context)
    {
        object callback=typeof(uiPortraitMessageHud).GetField("m_OnMessageClosed",Members).GetValue(context.portrait);
        foreach(string name in new[]{"DeliverQuestMessageClosedPart","DeliverStartQuestMsgPartClosed","DeliverMultiQuestMsgClosed","DeliverSubQuestMsgClosed"})
            if(StorySubmissionPolicy.ExactAction(callback,context.presenter,name))return name;
        return "unavailable";
    }
    StoryContext ReadStorySetup()
    {
        CatalogNoLinks(root);RequireSinglePlayer();
        JObject assembly=ScaleIdentity(typeof(MessageCoordinator).Assembly);
        if((string)assembly["assemblyFileSha256"]!=KrakenControllerGameAssembly)throw new InvalidOperationException("Story source audit requires exact game assembly.");
        StoryContext context=new StoryContext{coordinator=MessageCoordinator.Instance,presenter=MessagePresenter.Instance};
        FTKUI ui=FTKUI.Instance;GameLogic logic=GameLogic.Instance;GameFlow flow=GameFlow.Instance;FTKHub hub=FTKHub.Instance;
        if(context.coordinator==null || context.presenter==null || ui==null || logic==null || flow==null || hub==null)
            throw new InvalidOperationException("Native story coordinator/presenter/UI/world unavailable.");
        context.click=FTKClickAnywhere.Instance;context.portrait=ui.m_PortraitMessage;context.confirm=ui.m_QuestConfirm;context.cow=logic.GetCurrentCOW();
        MessageCoordinator.MessageInstance current=typeof(MessageCoordinator).GetField("m_CurrentMessageInstances",Members).GetValue(context.coordinator)as MessageCoordinator.MessageInstance;
        ICollection queue=typeof(MessageCoordinator).GetField("m_RequestQueue",Members).GetValue(context.coordinator)as ICollection;
        if(queue==null || queue.Count>64)throw new InvalidOperationException("Native story queue unavailable or outside bound64.");
        bool pendingContinuation=typeof(MessageCoordinator).GetField("m_CurrentCFSM",Members).GetValue(context.coordinator)!=null;
        QuestLogicBase quest=typeof(MessagePresenter).GetField("m_Quest",Members).GetValue(context.presenter)as QuestLogicBase;
        IDictionary questTable=typeof(GameLogic).GetProperty("_fullQuestTable",Members).GetValue(logic,null)as IDictionary;
        bool registeredQuest=quest!=null && StorySubmissionPolicy.RegisteredQuest(questTable,quest.m_QuestID,quest);
        QuestLogicBase confirmQuest=context.confirm==null?null:typeof(uiQuestConfirmHud).GetField("m_Quest",Members).GetValue(context.confirm)as QuestLogicBase;
        JObject portrait=StoryPanel(context.portrait,context.portrait==null?null:context.portrait.m_MessagePanel,
            context.portrait!=null && context.portrait.m_IsFullyOpened,context.portrait!=null && (bool)typeof(uiPortraitMessageHud).GetField("m_Clickable",Members).GetValue(context.portrait),context.portrait!=null && context.portrait.m_CloseOnOkay);
        JObject confirm=StoryPanel(context.confirm,context.confirm==null?null:context.confirm.m_MessagePanel,
            context.confirm!=null && context.confirm.m_IsFullyOpened,context.confirm!=null && (bool)typeof(uiQuestConfirmHud).GetField("m_ClickEnable",Members).GetValue(context.confirm),true);
        portrait["focus"]=StoryFocus(context.click,context.portrait);confirm["focus"]=StoryFocus(context.click,context.confirm);
        confirm["closeOnOkay"]=new JValue((object)null);
        confirm["samePresenterQuest"]=quest!=null && confirmQuest==quest;
        confirm["continuationPresent"]=context.confirm!=null && typeof(uiQuestConfirmHud).GetField("m_ContinueFSM",Members).GetValue(context.confirm)!=null;
        JObject bridge=JObject.FromObject(CatalogAssembly("FTKModFramework").GetType("FTKModFramework.Agent.StateReader",true).GetMethod("ReadState",Statics,null,Type.EmptyTypes,null).Invoke(null,null));
        bool outside=Instance(typeof(EncounterSession))!=null && Instance(typeof(EncounterSessionMC))!=null;try{RequireOutsideCombat();}catch(InvalidOperationException){outside=false;}
        bool liveParty=hub.m_CharacterOverworlds!=null && hub.m_CharacterOverworlds.Count>0,ownedCow=false;
        if(hub.m_CharacterOverworlds!=null)foreach(CharacterOverworld member in hub.m_CharacterOverworlds)
        {liveParty&=member!=null && member.m_CharacterStats!=null && member.m_CharacterStats.m_HealthCurrent>0;if(member==context.cow)ownedCow=true;}
        JArray fsms=new JArray{FsmView(typeof(MessagePresenter).GetField("m_MessageLoopFSM",Members).GetValue(context.presenter)),FsmView(typeof(MessagePresenter).GetField("m_StartQuestMsgLoopFSM",Members).GetValue(context.presenter)),FsmView(typeof(MessagePresenter).GetField("m_StartMultiQuestMsgLoopFSM",Members).GetValue(context.presenter))};
        string contentHash=StoryContentHash(context.presenter);
        context.view=new JObject{{"root",root},{"timeScale",Time.timeScale},{"provenance","read-only native story setup state"},{"coreIdentity",PreviewCoreIdentity()},{"gameAssembly",assembly},
            {"coordinatorId",context.coordinator.GetInstanceID()},{"messagePresent",current!=null},{"messageType",current==null?"None":current.m_MessageType.ToString()},
            {"messageId",current==null?-1:current.m_ID},{"messageClosed",current!=null && current.m_Closed},{"queueCount",queue.Count},{"currentContinuationPresent",pendingContinuation},
            {"presenterId",context.presenter.GetInstanceID()},{"presenterMessageId",context.presenter.m_MessageInstanceID},{"pageIndex",context.presenter.m_CurrentPageIndex},
            {"pageCount",context.presenter.m_Messages==null?0:context.presenter.m_Messages.Count},{"questId",quest==null?new JValue((object)null):new JValue(quest.m_QuestID)},{"questPresent",quest!=null},{"questRegisteredReference",registeredQuest},
            {"phase",context.portrait==null?"unavailable":StoryPhase(context)},{"contentAvailable",contentHash!=null},{"contentSha256",contentHash==null?new JValue((object)null):new JValue(contentHash)},
            {"presenterFsms",fsms},{"portrait",portrait},{"questConfirm",confirm},
            {"heroInstanceId",context.cow==null?0:context.cow.GetInstanceID()},{"livingOwnedParty",liveParty && ownedCow},
            {"outsideCombat",outside},{"outsideDungeon",flow.m_DungeonEntered==null && !hub.AnyPlayersInDungeon()},
            {"inSession",bridge["inSession"]},{"signals",bridge["signals"]},{"cameraWait",new JObject{{"available",false},{"note","No authoritative coroutine-wait flag. Page/FSM/panel observations describe pending work; do not infer a camera waiter."}}}};
        StorySubmissionPolicy.Classify(context.view);
        return context;
    }
    JObject StorySetupState(JObject command)
    {CatalogKeys(command,"id","session","op");JObject result=ReadStorySetup().view;result["ok"]=true;return result;}
    JObject SubmitStorySetup(JObject command)
    {
        CatalogKeys(command,"id","session","op","surface","coordinatorId","messageId","presenterId","pageIndex","questId","heroInstanceId","componentId","panelId","phase","contentSha256","clickInstanceId","clickContinuationId");
        StoryContext context=ReadStorySetup();JObject before=context.view;
        string key=StorySubmissionPolicy.Validate(command,before);
        if(submittedStoryPages.Count>=64 || !submittedStoryPages.Add(key))throw new InvalidOperationException("Story page already submitted or bounded64 submission limit reached; never retry uncertain actions.");
        // Record before the native call. An exception must never authorize another submission.
        context.click.OnClick();
        return new JObject{{"ok",true},{"status","native-story-page-submitted"},{"submissionKey",key},{"surface",command["surface"]},
            {"before",before},{"note","One native FTKClickAnywhere.OnClick call submitted to the exact UseOkayButton callback. Delayed closure/page advancement/completion is not yet observed; never retry this page."}};
    }
}
